pennylane/templates/tensornetworks/ttn.py

```diff
@@ -0,0 +1,210 @@
+        n_params_block (int): the number of parameters in a block; equal to the length of the ``weights`` argument in ``block``
+        template_weights (Sequence): list containing the weights for all blocks
+
+    .. note::
+
+        The expected number of blocks can be obtained from ``qml.TTN.n_blocks(wires, n_block_wires)``.
+        The length of ``template_weights`` argument should match the number of blocks.
+
+    .. UsageDetails::
+
+        This example demonstrates the use of ``TTN`` for a simple block.
+
+        .. code-block:: python
+
+            import pennylane as qml
+            import numpy as np
+
+            def block(weights, wires):
+                qml.CNOT(wires=[wires[0],wires[1]])
+                qml.RY(weights[0], wires=wires[0])
+                qml.RY(weights[1], wires=wires[1])
+
+            n_wires = 4
+            n_block_wires = 2
+            n_params_block = 2
+            n_blocks = qml.TTN.get_n_blocks(range(n_wires),n_block_wires)
+            template_weights = [[0.1,-0.3]]*n_blocks
+
+            dev= qml.device('default.qubit',wires=range(n_wires))
+            @qml.qnode(dev)
+            def circuit(template_weights):
+                qml.TTN(range(n_wires),n_block_wires,block, n_params_block, template_weights)
+                return qml.expval(qml.PauliZ(wires=n_wires-1))
+
+        >>> print(qml.draw(circuit,expansion_strategy='device')(template_weights))
+        0: ──╭C──RY(0.1)─────────────────┤
+        1: ──╰X──RY(-0.3)──╭C──RY(0.1)───┤
+        2: ──╭C──RY(0.1)───│─────────────┤
+        3: ──╰X──RY(-0.3)──╰X──RY(-0.3)──┤ ⟨Z⟩
+    """
+
+    num_params = 1
+    num_wires = AnyWires
+    par_domain = "A"
+
+    def __init__(
+        self,
+        wires,
+        n_block_wires,
+        block,
+        n_params_block,
+        template_weights=None,
+        do_queue=True,
+        id=None,
+    ):
+
+        self.ind_gates = compute_indices(wires, n_block_wires)
+        n_wires = len(wires)
+        shape = qml.math.shape(template_weights)[-4:]  # (n_params_block, n_blocks)
+        self.n_params_block = n_params_block
+        self.n_blocks = 2 ** int(np.log2(n_wires / n_block_wires)) * 2 - 1
+        self.block = block
+
+        if template_weights is None:
+            self.template_weights = np.random.rand(n_params_block, int(self.n_blocks))
+
+        else:
+
+            if shape[0] != self.n_blocks:
+                raise ValueError(
+                    f"Weights tensor must have first dimension of length {self.n_blocks}; got {shape[0]}"
+                )
+            if shape[-1] != self.n_params_block:
+                raise ValueError(
+                    f"Weights tensor must have last dimension of length {self.n_params_block}; got {shape[-1]}"
+                )
+
+            self.template_weights = template_weights
+
+        super().__init__(template_weights, wires=wires, do_queue=do_queue, id=id)
+
+    def expand(self):
```

Comment:
In the same refactor we start defining decompositions instead of expand, but I'll take care of this!